### PR TITLE
Ldap login scanner module

### LINKED ISSUE #18130: LDAP Login Scanner
Metasploit should have a LDAP login scanner module to allow users to test and bruteforce credentials against a target LDAP server. See the [docs](https://docs.metasploit.com/docs/development/developing-modules/guides/scanners/) on writing login scanners for some background information on how they are written. We have the capability to authenticate to LDAP already, just not a login scanner to test multiple username/password combinations.

### PR DESCRIPTION
**Pull request by dwelch-r7**
Adds a new login scanner module for LDAP
resolves #18130 

# Verification steps
- Run with valid and invalid creds for all authentication mechanisms
  - [x] Auto
  - [x] Plaintext
  - [x] NTLM
  - [x] Kerberos
    - [x] Logging in with kerberos successfully but without access to LDAP should not store the credentials
  - [x] SChannel 
- [x] Valid credentials are stored in the DB
- [x] Invalid credentials are not stored

#TODO: If the creds are used to get a kerberos ticket then that ticket is used for ldap access the creds are still stored in the db as plaintext password which is kind of confusing but I'm not sure what I'd expect the UX for that to be either

### CODE REVIEW

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
Some notes from attempting to use this at commit 6a673e to test #18125, shows behavior that suggests both PRs may need some adjustments.

```
[-] 192.168.222.49:389 - LOGIN FAILED: example.com\vagrant:vagrant (Incorrect: Bind Result: {:extended_response=>nil, :code=>7, :error_message=>"Auth Method Not Supported", :matched_dn=>"", :message=>"Auth Method Not Supported"})
```
Capture module ends up with
```
[+] LDAP Login Attempt => From:192.168.222.57:41883 Username: Password:
```
suggesting the auth portion is not getting submitted the as what ldapsearch such as:
```
ldapsearch -LLL -H ldap://192.168.222.49 -D admin@example.com -W
```
would result in:
```
[+] LDAP Login Attempt => From:192.168.222.57:44152 Username:admin Password:
```
and
```
Credentials
===========

host            origin          service         public  private  realm        private_type  
----            ------          -------         ------  -------  -----        ------------  
192.168.222.57  192.168.222.57  389/tcp (ldap)  admin   test     example.com  Password
```
showing the capture module is missing the password in logs but does capture `ldapsearch` based requests that use plaintext auth.

I am doing some further testing of the capture branch with other tools that should bruteforce as well so this might be partially on the framework listener but worth mentioning as y’all test the login module.

Some additional context setting the advanced LDAP:Auth option:
```
msf6 auxiliary(scanner/ldap/ldap_login) > set LDAP::Auth plaintext
LDAP::Auth => plaintext
msf6 auxiliary(scanner/ldap/ldap_login) > run

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
This pattern looks like it does not exercise the provided uses/pass list and if I add a `set DOMAIN` to this config I get a stacktrace:
```
msf6 auxiliary(scanner/ldap/ldap_login) > set DOMAIN example.com
DOMAIN => example.com
msf6 auxiliary(scanner/ldap/ldap_login) > run

[-] Auxiliary failed: NoMethodError undefined method `<<' for nil:NilClass
[-] Call stack:
[-]   /home/vagrant/rapid7/metasploit-framework/lib/metasploit/framework/ldap/client.rb:119:in `ldap_connect_opts'
[-]   /home/vagrant/rapid7/metasploit-framework/lib/msf/core/exploit/remote/ldap.rb:91:in `get_connect_opts'
[-]   /home/vagrant/rapid7/metasploit-framework/modules/auxiliary/scanner/ldap/ldap_login.rb:42:in `validate_connect_options!'
[-]   /home/vagrant/rapid7/metasploit-framework/modules/auxiliary/scanner/ldap/ldap_login.rb:35:in `run'
[*] Auxiliary module execution completed
```

File: lib/metasploit/framework/ldap/client.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            # username << "@#{opts[:domain]}" unless opts[:domain].blank?
+            connect_opts[:auth] = {
+              method: :simple,
+              username: opts[:username],
+              password: opts[:password]
+            }
+          when Msf::Exploit::Remote::AuthOption::AUTO
+            unless opts[:username].blank? # plaintext if specified
+              # username = opts[:username].dup
+              # username << "@#{opts[:domain]}" unless opts[:domain].blank?
```
Comment (adfoster-r7) [inline lib/metasploit/framework/ldap/client.rb:127]: Is this code meant to be commented out? 👀

Comment (dwelch-r7) [reply lib/metasploit/framework/ldap/client.rb:127]: I believe that's just left over from when I moved the code, should be able to remove it

**Review by jheysel-r7 — COMMENTED** [review]
I still have some testing to do, I'll hopefully wrap that up tomorrow. In the mean time, there's just one small fix that needs addressing.

**Review by jheysel-r7 — COMMENTED** [review]
Hey @dwelch-r7. I've done some testing and `auto`, `plaintext` and `ntlm` all seem to be working as expected. 

<details>
<summary>auto, plaintext and ntlm</summary>

<p>


```
msf6 auxiliary(scanner/ldap/ldap_login) > set LDAP::Auth auto
LDAP::Auth => auto
msf6 auxiliary(scanner/ldap/ldap_login) > set username test\\administrator
username => test\administrator
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] 192.168.123.188:389 - Success: 'test\administrator:notpassword'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpassword
password => incorrectpassword
msf6 auxiliary(scanner/ldap/ldap_login) > run

[-] 192.168.123.188:389 - LOGIN FAILED: test\administrator:incorrectpassword (Incorrect: Bind Result: {:extended_response=>nil, :code=>49, :error_message=>"80090308: LdapErr: DSID-0C0903D3, comment: AcceptSecurityContext error, data 52e, v3839\u0000", :matched_dn=>"", :message=>"Invalid Credentials"})
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set LDAP::Auth plaintext
LDAP::Auth => plaintext
msf6 auxiliary(scanner/ldap/ldap_login) > set password notpassword
password => notpassword
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] 192.168.123.188:389 - Success: 'test\administrator:notpassword'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpassword
password => incorrectpassword
msf6 auxiliary(scanner/ldap/ldap_login) > run

[-] 192.168.123.188:389 - LOGIN FAILED: test\administrator:incorrectpassword (Incorrect: Bind Result: {:extended_response=>nil, :code=>49, :error_message=>"80090308: LdapErr: DSID-0C0903D3, comment: AcceptSecurityContext error, data 52e, v3839\u0000", :matched_dn=>"", :message=>"Invalid Credentials"})
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set LDAP::Auth ntlm
LDAP::Auth => ntlm
msf6 auxiliary(scanner/ldap/ldap_login) > set password notpassword
password => notpassword
msf6 auxiliary(scanner/ldap/ldap_login) > set username administrator
username => administrator
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] 192.168.123.188:389 - Success: 'administrator:notpassword'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpassword
password => incorrectpassword
msf6 auxiliary(scanner/ldap/ldap_login) > run

[-] 192.168.123.188:389 - LOGIN FAILED: administrator:incorrectpassword (Incorrect: Bind Result: {:extended_response=>nil, :code=>49, :error_message=>"8009030C: LdapErr: DSID-0C090516, comment: AcceptSecurityContext error, data 52e, v3839\u0000", :matched_dn=>"", :message=>"Invalid Credentials"})
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host             origin           service         public              private      realm  private_type  JtR Format
----             ------           -------         ------              -------      -----  ------------  ----------
192.168.123.188  192.168.123.188  389/tcp (ldap)  test\administrator  notpassword         Password
192.168.123.188  192.168.123.188  389/tcp (ldap)  administrator       notpassword         Password

```

</p>

</details>


Both `kerberos` and `channel` are saving incorrect credentials to the database.  For `kerberos`, once you authenticate successfully with a valid password, after that you can set an incorrect password, the module will use the cached credentials authenticate successfully but then save in the incorrect creds to the database: 

<details>
<summary>kerberos</summary>

<p>

```
msf6 auxiliary(scanner/ldap/ldap_login) > run LDAP::Auth=kerberos LDAP::Rhostname=WIN-FDV8QFUMFP7.example.com DomainControllerRhost=172.16.199.235 

[+] 172.16.199.235:88 - Received a valid TGT-Response
[*] 172.16.199.235:389 - TGT MIT Credential Cache ticket saved to /home/msfuser/.msf4/loot/20230913103540_default_172.16.199.235_mit.kerberos.cca_138458.bin
[+] 172.16.199.235:88 - Received a valid TGS-Response
[*] 172.16.199.235:389 - TGS MIT Credential Cache ticket saved to /home/msfuser/.msf4/loot/20230913103540_default_172.16.199.235_mit.kerberos.cca_533257.bin
[+] 172.16.199.235:88 - Received a valid delegation TGS-Response
[+] 172.16.199.235:389 - Success: 'example.com\administrator:N0tpassword!'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpassword
password => incorrectpassword
msf6 auxiliary(scanner/ldap/ldap_login) > run

[*] 172.16.199.235:88 - Using cached credential for ldap/WIN-FDV8QFUMFP7.example.com@EXAMPLE.COM administrator@EXAMPLE.COM
[+] 172.16.199.235:389 - Success: 'example.com\administrator:incorrectpassword'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host            origin          service         public         private            realm        private_type  JtR Format
----            ------          -------         ------         -------            -----        ------------  ----------
172.16.199.235  172.16.199.235  389/tcp (ldap)  administrator  incorrectpassword  example.com  Password
172.16.199.235  172.16.199.235  389/tcp (ldap)  administrator  N0tpassword!       example.com  Password

```

</p>

</details>


For `schannel`, if you start with no loot or creds in the db, use `icpr_cert` to obtain a valid certificate, and then run `ldap_login` with the valid certificate but an invalid password the module reports success and stores the invalid password in the db:

<details>
<summary>schannel</summary
<p>

```
msf6 auxiliary(scanner/ldap/ldap_login) > creds --delete
Credentials
===========

host            origin          service         public                     private                                                                                              realm        private_type  JtR Format
----            ------          -------         ------                     -------                                                                                              -----        ------------  ----------
172.16.199.235  172.16.199.235  389/tcp (ldap)  administrator              incorrectpassword                                                                                    example.com  Password
172.16.199.235  172.16.199.235  445/tcp (smb)   Administrator@example.com  subject:/DC=com/DC=example/CN=Users/CN=Administrator,issuer:/DC=com/DC=example/CN=examp (TRUNCATED)               Pkcs12 (pfx)

[*] Deleted 2 creds
msf6 auxiliary(scanner/ldap/ldap_login) > loot --delete

Loot
====

host            service  type           name             content               info                               path
----            -------  ----           ----             -------               ----                               ----
172.16.199.235           windows.ad.cs  certificate.pfx  application/x-pkcs12  EXAMPLE\Administrator Certificate  /home/msfuser/.msf4/loot/20230913112313_default_172.16.199.235_windows.ad.cs_731632.pfx

[*] Deleted 1 loots
msf6 auxiliary(scanner/ldap/ldap_login) > previous
msf6 auxiliary(admin/dcerpc/icpr_cert) > run
[*] Running module against 172.16.199.235

[*] 172.16.199.235:445 - Connecting to ICertPassage (ICPR) Remote Protocol
[*] 172.16.199.235:445 - Binding to \cert...
[+] 172.16.199.235:445 - Bound to \cert
[*] 172.16.199.235:445 - Requesting a certificate for user Administrator - digest algorithm: SHA256 - template: User
[+] 172.16.199.235:445 - The requested certificate was issued.
[*] 172.16.199.235:445 - Certificate UPN: Administrator@example.com
[*] 172.16.199.235:445 - Certificate stored at: /home/msfuser/.msf4/loot/20230913112401_default_172.16.199.235_windows.ad.cs_626130.pfx
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/icpr_cert) > previous
msf6 auxiliary(scanner/ldap/ldap_login) > run LDAP::Auth=schannel LDAP::CertFile=/home/msfuser/.msf4/loot/20230913112401_default_172.16.199.235_windows.ad.cs_626130.pfx ssl=true

[+] 172.16.199.235:389 - Success: 'administrator:incorrectpassword'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host            origin          service         public                     private                                                                                              realm  private_type  JtR Format
----            ------          -------         ------                     -------                                                                                              -----  ------------  ----------
172.16.199.235  172.16.199.235  389/tcp (ldap)  administrator              incorrectpassword                                                                                           Password
172.16.199.235  172.16.199.235  445/tcp (smb)   Administrator@example.com  subject:/DC=com/DC=example/CN=Users/CN=Administrator,issuer:/DC=com/DC=example/CN=examp (TRUNCATED)         Pkcs12 (pfx)
```

</p>

</details>

**Comment by dwelch-r7** [comment]
> Hey @dwelch-r7. I've done some testing and `auto`, `plaintext` and `ntlm` all seem to be working as expected.
> 
> auto, plaintext and ntlm
> Both `kerberos` and `channel` are saving incorrect credentials to the database. For `kerberos`, once you authenticate successfully with a valid password, after that you can set an incorrect password, the module will use the cached credentials authenticate successfully but then save in the incorrect creds to the database:
> 
> kerberos
> For `schannel`, if you start with no loot or creds in the db, use `icpr_cert` to obtain a valid certificate, and then run `ldap_login` with the valid certificate but an invalid password the module reports success and stores the invalid password in the db:
> 
> schannel

Thanks for testing, I think I've got changes up now that should address them all
turns out the smb_login module also has some of the same issues, I haven't fixed them all here though, if we're happy with what I've done here though I'll fire up another PR to address them there

**Comment by jheysel-r7** [comment]
Just tested the latest changes. Kerberos testing is behaving as expected now: 
```
msf6 auxiliary(scanner/ldap/ldap_login) > run LDAP::Auth=kerberos LDAP::Rhostname=WIN-FDV8QFUMFP7.example.com DomainControllerRhost=172.16.199.235 rhosts=172.16.199.235 domain=example.com username=administrator password=N0tpassword!

[+] 172.16.199.235:88 - Received a valid TGT-Response
[*] 172.16.199.235:389 - TGT MIT Credential Cache ticket saved to /home/msfuser/.msf4/loot/20230929123924_default_172.16.199.235_mit.kerberos.cca_794982.bin
[+] 172.16.199.235:88 - Received a valid TGS-Response
[*] 172.16.199.235:389 - TGS MIT Credential Cache ticket saved to /home/msfuser/.msf4/loot/20230929123924_default_172.16.199.235_mit.kerberos.cca_652233.bin
[+] 172.16.199.235:88 - Received a valid delegation TGS-Response
[+] 172.16.199.235:389 - Success: 'example.com\administrator:N0tpassword!'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > run LDAP::Auth=kerberos LDAP::Rhostname=WIN-FDV8QFUMFP7.example.com DomainControllerRhost=172.16.199.235 rhosts=172.16.199.235 domain=example.com username=administrator password=incorrectpassword

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host            origin          service         public         private       realm        private_type  JtR Format
----            ------          -------         ------         -------       -----        ------------  ----------
172.16.199.235  172.16.199.235  389/tcp (ldap)  administrator  N0tpassword!  example.com  Password
```

However schannel still seems to be reporting auth attempts with incorrect creds as successful as well as saving those incorrect creds in the DB - @dwelch-r7 is this issue one of the same issues the smb_login module was facing? Let me know if you think this issue would be better fixed as apart of another PR :
```
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] 172.16.199.132:389 - Success: 'Administrator:N0tpassword!'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpass
password => incorrectpass
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] 172.16.199.132:389 - Success: 'Administrator:incorrectpass'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host            origin          service         public         private        realm  private_type  JtR Format
----            ------          -------         ------         -------        -----  ------------  ----------
172.16.199.132  172.16.199.132  389/tcp (ldap)  Administrator  incorrectpass         Password
172.16.199.132  172.16.199.132  389/tcp (ldap)  Administrator  N0tpassword!          Password

```

<details>

<summary>schannel module run - show advanced options </summary>

<p>

```

 msf6 auxiliary(scanner/ldap/ldap_login) > show advanced options

Module advanced options (auxiliary/scanner/ldap/ldap_login):

   Name                  Current Setting  Required  Description
   ----                  ---------------  --------  -----------
   LDAP::Auth            schannel         yes       The Authentication mechanism to use (Accepted: auto, ntlm, kerberos, schannel, plaintext, none)
   LDAP::ConnectTimeout  10.0             yes       Timeout for LDAP connect
   MaxGuessesPerService  0                no        Maximum number of credentials to try per service instance. If set to zero or a non-number, this option will not be used.
   MaxGuessesPerUser     0                no        Maximum guesses for a particular username for the service instance. Note that users are considered unique among different services, so a user at 10.1.1.1:22 is different from one
                                                    at 10.2.2.2:22, and both will be tried up to the MaxGuessesPerUser limit. If set to zero or a non-number, this option will not be used.
   MaxMinutesPerService  0                no        Maximum time in minutes to bruteforce the service instance. If set to zero or a non-number, this option will not be used.
   PASSWORD_SPRAY        false            yes       Reverse the credential pairing order. For each password, attempt every possible user.
   Proxies                                no        A proxy chain of format type:host:port[,type:host:port][...]
   REMOVE_PASS_FILE      false            yes       Automatically delete the PASS_FILE on module completion
   REMOVE_USERPASS_FILE  false            yes       Automatically delete the USERPASS_FILE on module completion
   REMOVE_USER_FILE      false            yes       Automatically delete the USER_FILE on module completion
   ShowProgress          true             yes       Display progress messages during a scan
   ShowProgressPercent   10               yes       The interval in percent that progress should be shown
   TRANSITION_DELAY      0                no        Amount of time (in minutes) to delay before transitioning to the next user in the array (or password when PASSWORD_SPRAY=true)
   WORKSPACE                              no        Specify the workspace for this module


   Active when LDAP::Auth is kerberos:

   Name                             Current Setting                                   Required  Description
   ----                             ---------------                                   --------  -----------
   DomainControllerRhost                                                              no        The resolvable rhost for the Domain Controller
   KrbCacheMode                     read-write                                        yes       Kerberos ticket cache storage mode (Accepted: none, read-only, write-only, read-write)
   LDAP::Krb5Ccname                                                                   no        The ccache file to use for kerberos authentication
   LDAP::KrbOfferedEncryptionTypes  AES256,AES128,RC4-HMAC,DES-CBC-MD5,DES3-CBC-SHA1  yes       Kerberos encryption types to offer
   LDAP::Rhostname                  WIN-FDV8QFUMFP7.example.com                       no        The rhostname which is required for kerberos - the SPN


   Active when LDAP::Auth is schannel:

   Name            Current Setting                                                                          Required  Description
   ----            ---------------                                                                          --------  -----------
   LDAP::CertFile  /home/msfuser/.msf4/loot/20230929142723_default_172.16.199.132_windows.ad.cs_517539.pfx  no        The path to the PKCS12 (.pfx) certificate file to authenticate with


View the full module info with the info, or info -d command.


Module options (auxiliary/scanner/ldap/ldap_login):

   Name              Current Setting  Required  Description
   ----              ---------------  --------  -----------
   ANONYMOUS_LOGIN   false            yes       Attempt to login with a blank username and password
   BRUTEFORCE_SPEED  5                yes       How fast to bruteforce, from 0 to 5
   DB_ALL_CREDS      false            no        Try each user/password couple stored in the current database
   DB_ALL_PASS       false            no        Add all passwords in the current database to the list
   DB_ALL_USERS      false            no        Add all users in the current database to the list
   DB_SKIP_EXISTING  none             no        Skip existing credentials stored in the current database (Accepted: none, user, user&realm)
   DOMAIN            example.com      no        The domain to authenticate to
   PASSWORD          incorrectpass    no        The password to authenticate with
   PASS_FILE                          no        File containing passwords, one per line
   RHOSTS            172.16.199.132   yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT             389              yes       The target port
   SSL               true             no        Enable SSL on the LDAP connection
   STOP_ON_SUCCESS   false            yes       Stop guessing when a credential works for a host
   THREADS           1                yes       The number of concurrent threads (max one per host)
   USERNAME          Administrator    no        The username to authenticate with
   USERPASS_FILE                      no        File containing users and passwords separated by space, one pair per line
   USER_AS_PASS      false            no        Try the username as the password for all users
   USER_FILE                          no        File containing usernames, one per line
   VERBOSE           true             yes       Whether to print output for all attempts


   When LDAP::Auth is one of auto,plaintext:

   Name           Current Setting  Required  Description
   ----           ---------------  --------  -----------
   APPEND_DOMAIN  false            yes       Appends `@<DOMAIN> to the username for authentication`


View the full module info with the info, or info -d command.

```

</p> 

</details>

**Comment by dwelch-r7** [comment]
Sorry about that @jheysel-r7 I was still storing the username for schannel but it's also not used in that auth workflow, this this did have the side effect of not printing anything when you were successful  with the schannel auth so I special cased it
We _can_ store pkcs12 in the db but since you can only use schannel auth by passing it a path to the file you already have saving it seemed a bit weird and we don't do the same thing when you give framework a ccache file not tracked in the db

**Comment by jheysel-r7** [comment]
Thanks @dwelch-r7!! Testing checks out. Everything is looking good, I'll go ahead and land this. 

```
msf6 auxiliary(scanner/ldap/ldap_login) > set password incorrectpass
password => incorrectpass
msf6 auxiliary(scanner/ldap/ldap_login) > run

[+] Success: 'Cert File /home/msfuser/.msf4/loot/20231002064918_default_172.16.199.132_windows.ad.cs_263738.pfx'
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ldap/ldap_login) > creds
Credentials
===========

host  origin  service  public  private  realm  private_type  JtR Format
----  ------  -------  ------  -------  -----  ------------  ----------

msf6 auxiliary(scanner/ldap/ldap_login) >
```

**Pull request merged**